File: SGUEES-SPA/src/app/pages/Academics/aca-prospecto/aca-prospecto-estudio/aca-prospecto-estudio.service.ts
import { Observable } from 'rxjs';
import { Injectable } from '@angular/core';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';

import { AcaProspectoEstudioRepository } from './aca-prospecto-estudio.repository';

// Qué hace: secciones de estudios que maneja el prospecto (las mismas que arma la vista).
// Cómo lo hace: el portal guarda como máximo una fila por sección (índice único por persona,
//               nivel y graduado UEES), así que cada sección es un formulario, no una grilla.
export const SECCION_MEDIA = 'MEDIA';
export const SECCION_UNIVERSIDAD = 'UNIVERSIDAD';
export const SECCION_GRADUADO_UEES = 'GRADUADO_UEES';

@Injectable({
    providedIn: 'root',
})
export class AcaProspectoEstudioService {
    constructor(private repo: AcaProspectoEstudioRepository) {}

    // Qué hace: estudios previos del prospecto (pestaña Información académica).
    // Cómo lo hace: la API los devuelve ordenados por sección (media, universidad, graduado UEES).
    getEstudiosPorProspecto(CORR_PROSPECTO: number): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO', Value: CORR_PROSPECTO }];
        return this.repo.getAll(xWhere);
    }

    // Qué hace: modelo vacío de una sección, listo para crearse al guardar.
    // Cómo lo hace: fija el nivel y la bandera de graduado UEES, que son los que definen la sección;
    //               el graduado UEES siempre es de la institución UEES y cuenta como graduado.
    getModeloNuevo(seccion: string, persona: any): any {
        const esUees = seccion === SECCION_GRADUADO_UEES;
        return {
            CORR_PROSPECTO_ESTUDIO: 0,
            CORR_PROSPECTO_PERSONA: persona?.CORR_PROSPECTO_PERSONA ?? 0,
            CORR_PROSPECTO: persona?.CORR_PROSPECTO ?? 0,
            SECCION: seccion,
            NIVEL_ESTUDIO: seccion === SECCION_MEDIA ? 'MEDIO' : 'SUPERIOR',
            GRADUADO_UEES: esUees,
            GRADUADO: seccion !== SECCION_UNIVERSIDAD,
            NOMBRE_INSTITUCION: esUees ? 'UEES' : '',
            TIPO_EDUCACION: null,
            TITULO_OBTENIDO: '',
            CARRERA_TEXTO: '',
            NIVEL_CURSADO: '',
            CORR_CARRERA: null,
            ANIO_TITULACION: null,
            FECHA_GRADUACION: null,
            CUOTA: null,
            CORR_PAIS: null,
            CORR_DEPTO: null,
        };
    }

    insert(model: any): Observable<IResult> {
        return this.repo.create(this.getPayload(model));
    }

    update(model: any): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO_ESTUDIO', Value: model.CORR_PROSPECTO_ESTUDIO }];
        return this.repo.update(this.getPayload(model), xWhere);
    }

    delete(model: any): Observable<IResult> {
        let xWhere: IParam[] = [{ Parameter: 'CORR_PROSPECTO_ESTUDIO', Value: model.CORR_PROSPECTO_ESTUDIO }];
        return this.repo.delete(xWhere);
    }

    // Qué hace: formulario "Educación media (bachillerato)".
    // Cómo lo hace: mismos campos y obligatoriedades del portal; país y departamento son lookups
    //               (template), así que su obligatoriedad se revisa en validar().
    getItemsMedia(): any {
        return [
            {
                dataField: 'NOMBRE_INSTITUCION',
                label: { text: 'Nombre de la institución' },
                colSpan: 4,
                editorOptions: { maxLength: 100 },
                validationRules: [{ type: 'required', message: 'Ingrese la institución' }],
            },
            {
                dataField: 'TIPO_EDUCACION',
                label: { text: 'Tipo de educación' },
                colSpan: 2,
                editorType: 'dxSelectBox',
                editorOptions: { items: this.getTiposEducacion(), showClearButton: true },
                validationRules: [{ type: 'required', message: 'Seleccione el tipo de educación' }],
            },
            { dataField: 'CORR_PAIS', label: { text: 'País' }, colSpan: 2, template: 'CORR_PAIS_ESTUDIOLookup' },
            { dataField: 'CORR_DEPTO', label: { text: 'Departamento' }, colSpan: 2, template: 'CORR_DEPTO_ESTUDIOLookup' },
            {
                dataField: 'TITULO_OBTENIDO',
                label: { text: 'Título obtenido' },
                colSpan: 4,
                editorOptions: { maxLength: 200 },
                validationRules: [{ type: 'required', message: 'Ingrese el título obtenido' }],
            },
            {
                dataField: 'ANIO_TITULACION',
                label: { text: 'Año de titulación' },
                colSpan: 2,
                editorType: 'dxNumberBox',
                editorOptions: { format: '#', min: 1900, max: new Date().getFullYear(), showClearButton: true },
                validationRules: [{ type: 'required', message: 'Indique el año de titulación' }],
            },
            {
                dataField: 'CUOTA',
                label: { text: 'Cuota ($)' },
                colSpan: 2,
                editorType: 'dxNumberBox',
                editorOptions: { format: '#,##0.00', min: 0, showClearButton: true },
            },
            // El NIE se pregunta aquí, igual que en el portal, aunque la columna sea de la persona:
            // por eso va como template (lo edita el componente sobre personaModel), no como dataField.
            {
                name: 'NIE_PERSONA',
                label: { text: 'NIE' },
                colSpan: 2,
                template: 'NIE_PERSONATemplate',
                helpText: 'Número de identificación del estudiante',
            },
        ];
    }

    // Qué hace: formulario "Estudios universitarios" (universidad de procedencia, distinta de la UEES).
    // Cómo lo hace: replica la regla del portal: los datos solo se llenan si el ingreso no es nuevo
    //               ingreso o si el prospecto es graduado universitario; si no, quedan bloqueados y
    //               sin exigencias (el portal guarda la fila con la institución vacía). El título y la
    //               cuota se exigen cuando está graduado. "¿Se graduó?" siempre se puede cambiar:
    //               es parte de la condición que habilita el resto.
    getItemsUniversidad(graduado = false, habilitado = true): any {
        const requerido = (mensaje: string) => (habilitado ? [{ type: 'required', message: mensaje }] : undefined);
        return [
            {
                dataField: 'GRADUADO',
                label: { text: '¿Es graduado universitario?' },
                colSpan: 2,
                editorType: 'dxCheckBox',
            },
            {
                dataField: 'NOMBRE_INSTITUCION',
                label: { text: 'Universidad de procedencia' },
                colSpan: 4,
                editorOptions: { maxLength: 100, readOnly: !habilitado },
                validationRules: requerido('Ingrese la universidad de procedencia'),
            },
            {
                dataField: 'CARRERA_TEXTO',
                label: { text: 'Carrera' },
                colSpan: 2,
                editorOptions: { maxLength: 100, readOnly: !habilitado },
            },
            {
                dataField: 'NIVEL_CURSADO',
                label: { text: 'Nivel cursado' },
                colSpan: 2,
                editorOptions: { maxLength: 10, readOnly: !habilitado },
                validationRules: requerido('Indique el nivel cursado'),
            },
            {
                dataField: 'TITULO_OBTENIDO',
                label: { text: 'Título obtenido' },
                colSpan: 4,
                editorOptions: { maxLength: 200, readOnly: !habilitado },
                validationRules: habilitado && graduado ? [{ type: 'required', message: 'Ingrese el título obtenido' }] : undefined,
            },
            {
                dataField: 'CUOTA',
                label: { text: 'Cuota ($)' },
                colSpan: 2,
                editorType: 'dxNumberBox',
                editorOptions: { format: '#,##0.00', min: 0, showClearButton: true, readOnly: !habilitado },
                validationRules: habilitado && graduado ? [{ type: 'required', message: 'Indique la cuota' }] : undefined,
            },
        ];
    }

    // Qué hace: dice si se pueden llenar los datos de la universidad de procedencia.
    // Cómo lo hace: misma condición del portal (universityGraduated = 'N' y typeAdmission = 'NI' los
    //               deshabilita): se habilitan si el ingreso no es nuevo ingreso o si es graduado.
    universidadHabilitada(formaIngreso: string, graduado: boolean): boolean {
        return (formaIngreso ?? '') !== 'NI' || graduado === true;
    }

    // Qué hace: formulario "Graduado UEES" (lo mismo que pide el portal: carrera y fecha).
    // Cómo lo hace: el grado académico no se muestra ni se captura; el API lo deriva de la carrera.
    getItemsUees(): any {
        return [
            { dataField: 'CORR_CARRERA', label: { text: 'Carrera' }, colSpan: 4, template: 'CORR_CARRERA_ESTUDIOLookup' },
            {
                dataField: 'FECHA_GRADUACION',
                label: { text: 'Fecha de graduación' },
                colSpan: 2,
                editorType: 'dxDateBox',
                editorOptions: { type: 'date', displayFormat: 'dd/MM/yyyy', dateSerializationFormat: 'yyyy-MM-dd' },
                validationRules: [{ type: 'required', message: 'Indique la fecha de graduación' }],
            },
        ];
    }

    // Qué hace: valores de "Tipo de educación".
    // Cómo lo hace: no es catálogo de base de datos; el portal los tiene fijos en su procedimiento.
    getTiposEducacion(): string[] {
        return ['Pública', 'Privada'];
    }

    // Qué hace: obligatoriedades que el dx-form no puede validar (los catálogos se pintan con template).
    // Cómo lo hace: devuelve el mensaje a mostrar o null si la sección está completa.
    validar(model: any): string | null {
        if (model?.SECCION === SECCION_MEDIA) {
            if (!model?.CORR_PAIS) {
                return 'Seleccione el país de la institución de educación media.';
            }
            if (!model?.CORR_DEPTO) {
                return 'Seleccione el departamento de la institución de educación media.';
            }
            return null;
        }
        if (model?.SECCION === SECCION_GRADUADO_UEES && !model?.CORR_CARRERA) {
            return 'Seleccione la carrera con la que se graduó en la UEES.';
        }
        return null;
    }

    // Qué hace: columnas de ACA_PROSPECTO_ESTUDIO que viajan al API.
    // Cómo lo hace: manda solo lo que tiene sentido en cada sección; el API vuelve a limpiar el resto
    //               y deriva el grado académico cuando es graduado UEES. Es público porque el
    //               componente lo usa además para saber si la sección cambió.
    getPayload(model: any): any {
        const esMedia = model.SECCION === SECCION_MEDIA;
        const esUees = model.SECCION === SECCION_GRADUADO_UEES;
        return {
            CORR_PROSPECTO_ESTUDIO: model.CORR_PROSPECTO_ESTUDIO ?? 0,
            CORR_PROSPECTO_PERSONA: model.CORR_PROSPECTO_PERSONA,
            NIVEL_ESTUDIO: esMedia ? 'MEDIO' : 'SUPERIOR',
            GRADUADO_UEES: esUees,
            GRADUADO: esUees || esMedia ? true : model.GRADUADO === true,
            // La columna es NOT NULL: el portal guarda vacío cuando el bloque no aplica.
            NOMBRE_INSTITUCION: esUees ? 'UEES' : model.NOMBRE_INSTITUCION ?? '',
            TIPO_EDUCACION: esMedia ? model.TIPO_EDUCACION : null,
            TITULO_OBTENIDO: esUees ? null : model.TITULO_OBTENIDO,
            CARRERA_TEXTO: esMedia || esUees ? null : model.CARRERA_TEXTO,
            NIVEL_CURSADO: esMedia || esUees ? null : model.NIVEL_CURSADO,
            CORR_CARRERA: esUees ? model.CORR_CARRERA : null,
            ANIO_TITULACION: esMedia ? model.ANIO_TITULACION : null,
            FECHA_GRADUACION: esUees ? model.FECHA_GRADUACION : null,
            CUOTA: esUees ? null : model.CUOTA,
            QUIEN_PAGO_CUOTA: model.QUIEN_PAGO_CUOTA,
            CORR_PAIS: esMedia ? model.CORR_PAIS : null,
            CORR_DEPTO: esMedia ? model.CORR_DEPTO : null,
            CORR_MUNICIPIO: null,
        };
    }
}

File: SGUEES-API/sguees.api/Options/Academics/ACA_PROSPECTO_ESTUDIO/ACA_PROSPECTO_ESTUDIOService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using sguees.Models;
using sguees.Repositories;

namespace sguees.Services
{
    public class ACA_PROSPECTO_ESTUDIOService : IACA_PROSPECTO_ESTUDIOService
    {
        private readonly IACA_PROSPECTO_ESTUDIORepository _repo;

        public ACA_PROSPECTO_ESTUDIOService(IACA_PROSPECTO_ESTUDIORepository repo)
        {
            _repo = repo;
        }

        // Qué hace: estudios previos del prospecto (pestaña Información académica).
        // Cómo lo hace: exige CORR_PROSPECTO porque eFramework omite del WHERE los enteros en 0
        //               y, sin él, devolvería los datos de todos los prospectos.
        public async Task<CResult> GetAllAsync(ACA_PROSPECTO_ESTUDIOParam xWhere)
        {
            if (xWhere.CORR_PROSPECTO <= 0)
                return new CResult() { Data = null, Result = false, ErrorCode = -1, ErrorMessage = "Debe indicar el prospecto" };

            var p = new List<CParameter>
            {
                new CParameter() {ParameterName="CORR_EMPRESA",Value=xWhere.CORR_EMPRESA,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO",Value=xWhere.CORR_PROSPECTO,DbType=System.Data.DbType.Int32},
            };

            return await _repo.GetAllAsync(p);
        }

        // Qué hace: un registro por su llave (CORR_PROSPECTO_ESTUDIO); sin llave, el primero del prospecto.
        // Cómo lo hace: exige CORR_PROSPECTO por la misma razón que GetAllAsync;
        //               la llave es opcional porque eFramework omite los enteros en 0.
        public async Task<CResult> GetAsync(ACA_PROSPECTO_ESTUDIOParam xWhere)
        {
            if (xWhere.CORR_PROSPECTO <= 0)
                return new CResult() { Data = null, Result = false, ErrorCode = -1, ErrorMessage = "Debe indicar el prospecto" };

            var p = new List<CParameter>
            {
                new CParameter() {ParameterName="CORR_EMPRESA",Value=xWhere.CORR_EMPRESA,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO",Value=xWhere.CORR_PROSPECTO,DbType=System.Data.DbType.Int32},
                new CParameter() {ParameterName="CORR_PROSPECTO_ESTUDIO",Value=xWhere.CORR_PROSPECTO_ESTUDIO,DbType=System.Data.DbType.Int32},
            };

            return await _repo.GetAsync(p);
        }

        // Qué hace: valida y crea un estudio del prospecto.
        // Cómo lo hace: mismas reglas que la modificación más la persona obligatoria; el repositorio
        //               rechaza un segundo estudio de la misma sección (índice único de la tabla).
        public async Task<CResult> CreateAsync(ACA_PROSPECTO_ESTUDIOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_PERSONA <= 0)
                return ErrorValidacion("Debe indicar la persona del prospecto.");

            var validacion = Validar(Data);
            if (validacion != null) return validacion;

            return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        public async Task<CResult> UpdateAsync(ACA_PROSPECTO_ESTUDIOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_ESTUDIO <= 0)
                return ErrorValidacion("Debe indicar el estudio a modificar.");

            var validacion = Validar(Data);
            if (validacion != null) return validacion;

            return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        public async Task<CResult> DeleteAsync(ACA_PROSPECTO_ESTUDIOTable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            if (Data == null || Data.CORR_PROSPECTO_ESTUDIO <= 0)
                return ErrorValidacion("Debe indicar el estudio a eliminar.");

            return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: reglas de alta y modificación, iguales a las que aplica el portal de admisiones.
        // Cómo lo hace: la sección se deduce de NIVEL_ESTUDIO + GRADUADO_UEES y de ahí salen los campos
        //               con sentido; los del resto de secciones se limpian para que no queden datos
        //               cruzados (el formulario de cada bloque tampoco los muestra).
        private static CResult Validar(ACA_PROSPECTO_ESTUDIOTable Data)
        {
            Data.NIVEL_ESTUDIO = (Limpiar(Data.NIVEL_ESTUDIO) ?? "").ToUpperInvariant();
            if (Data.NIVEL_ESTUDIO != NivelMedio && Data.NIVEL_ESTUDIO != NivelSuperior)
                return ErrorValidacion("El nivel de estudio debe ser MEDIO o SUPERIOR.");

            Data.NOMBRE_INSTITUCION = Limpiar(Data.NOMBRE_INSTITUCION);
            Data.TITULO_OBTENIDO = Limpiar(Data.TITULO_OBTENIDO);
            Data.CARRERA_TEXTO = Limpiar(Data.CARRERA_TEXTO);
            Data.NIVEL_CURSADO = Limpiar(Data.NIVEL_CURSADO);
            Data.TIPO_EDUCACION = Limpiar(Data.TIPO_EDUCACION);
            Data.QUIEN_PAGO_CUOTA = Limpiar(Data.QUIEN_PAGO_CUOTA);
            Data.BACHILLER_OPCION = Limpiar(Data.BACHILLER_OPCION);

            if (Data.NIVEL_ESTUDIO == NivelMedio)
            {
                Data.GRADUADO_UEES = false;
                Data.GRADUADO = true;
                Data.CORR_CARRERA = null;
                Data.CARRERA_TEXTO = null;
                Data.NIVEL_CURSADO = null;
                Data.FECHA_GRADUACION = null;
            }
            else if (Data.GRADUADO_UEES)
            {
                // Graduado UEES: la institución es fija y el grado lo deriva el repositorio de la carrera.
                Data.NOMBRE_INSTITUCION = "UEES";
                Data.GRADUADO = true;
                Data.TIPO_EDUCACION = null;
                Data.CARRERA_TEXTO = null;
                Data.NIVEL_CURSADO = null;
                Data.ANIO_TITULACION = null;
                Data.CORR_PAIS = null;
                Data.CORR_DEPTO = null;
                Data.CORR_MUNICIPIO = null;

                if (!(Data.CORR_CARRERA > 0))
                    return ErrorValidacion("Seleccione la carrera con la que se graduó en la UEES.");
                if (Data.FECHA_GRADUACION == null)
                    return ErrorValidacion("Indique la fecha de graduación.");
            }
            else
            {
                // Universidad de procedencia: el portal solo guarda texto libre, sin carrera del catálogo.
                // La institución puede ir vacía: con nuevo ingreso y sin ser graduado universitario, el
                // portal deshabilita esos campos y guarda la fila en blanco (la columna es NOT NULL).
                Data.NOMBRE_INSTITUCION ??= string.Empty;
                Data.CORR_CARRERA = null;
                Data.CORR_GRADO_ACADEMICO = null;
                Data.TIPO_EDUCACION = null;
                Data.ANIO_TITULACION = null;
                Data.FECHA_GRADUACION = null;
                Data.CORR_PAIS = null;
                Data.CORR_DEPTO = null;
                Data.CORR_MUNICIPIO = null;
            }

            if (Data.NOMBRE_INSTITUCION == null)
                return ErrorValidacion("Ingrese el nombre de la institución.");

            var largo = ExcedeLargo(Data.NOMBRE_INSTITUCION, 100, "El nombre de la institución")
                ?? ExcedeLargo(Data.TITULO_OBTENIDO, 200, "El título obtenido")
                ?? ExcedeLargo(Data.CARRERA_TEXTO, 100, "La carrera")
                ?? ExcedeLargo(Data.NIVEL_CURSADO, 10, "El nivel cursado")
                ?? ExcedeLargo(Data.TIPO_EDUCACION, 20, "El tipo de educación")
                ?? ExcedeLargo(Data.QUIEN_PAGO_CUOTA, 200, "Quién pagó la cuota")
                ?? ExcedeLargo(Data.BACHILLER_OPCION, 100, "La opción de bachillerato");
            if (largo != null) return largo;

            if (Data.CUOTA < 0)
                return ErrorValidacion("La cuota no puede ser negativa.");
            if (Data.ANIO_TITULACION != null && (Data.ANIO_TITULACION < 1900 || Data.ANIO_TITULACION > DateTime.Now.Year))
                return ErrorValidacion($"El año de titulación debe estar entre 1900 y {DateTime.Now.Year}.");
            if (Data.CORR_DEPTO > 0 && !(Data.CORR_PAIS > 0))
                return ErrorValidacion("Debe seleccionar el país antes del departamento.");
            if (Data.CORR_MUNICIPIO > 0 && !(Data.CORR_DEPTO > 0))
                return ErrorValidacion("Debe seleccionar el departamento antes del municipio.");

            return null;
        }

        private const string NivelMedio = "MEDIO";
        private const string NivelSuperior = "SUPERIOR";

        private static string Limpiar(string valor) => string.IsNullOrWhiteSpace(valor) ? null : valor.Trim();

        private static CResult ExcedeLargo(string valor, int maximo, string campo)
        {
            return valor != null && valor.Length > maximo ? ErrorValidacion($"{campo} no puede superar {maximo} caracteres.") : null;
        }

        private static CResult ErrorValidacion(string mensaje)
        {
            return new CResult() { Data = null, Result = false, CodeHelper = 0, ErrorCode = -1, ErrorMessage = mensaje, ErrorSource = "[ACA_PROSPECTO_ESTUDIOService]", RowsAffected = 0 };
        }
    }
}
